internal/extensions/extensions.go

```diff
@@ -0,0 +1,219 @@
+	return fmt.Sprintf("Internal extension format is invalid (%v). It should be in the format `internal:<extension>` with only lowercase letters and underscores", e.invalidExt)
+}
+
+type invalidExternalExtensionErr struct {
+	invalidExt string
+}
+
+func (e *invalidExternalExtensionErr) Error() string {
+	return fmt.Sprintf("External extension format is invalid (%v). It should be in the format `<extension>` with only lowercase letters and underscores", e.invalidExt)
+}
+
+// Extensions represents a registry of extensions.
+type Extensions struct {
+	// internal extensions are related to MicroCluster capabilities, not capabilities of
+	// MicroCluster-backed services.
+	//
+	// To distinguish between internal and external extensions, internal extensions
+	// are prefixed with a `internal:` value.
+	// e.g, `internal:runtime_extension_v1`, `internal:runtime_extension_v2`, ...
+	internal map[string]struct{}
+
+	// External extensions are related to capabilities of MicroCluster-backed services.
+	// e.g, `microovn_custom_encapsulation_ip`
+	external map[string]struct{}
+}
+
+// validateExternalExtension validates the given external extension.
+func validateExternalExtension(extension string) error {
+	if extension == "" {
+		return fmt.Errorf("Extension cannot be empty")
+	}
+
+	// Aside from the internal extensions, we let the extension to be like
+	// `<extension>` with only lowercase letters and underscores.
+	if !externalExtensionRegex.MatchString(extension) {
+		return &invalidExternalExtensionErr{invalidExt: extension}
+	}
+
+	return nil
+}
+
+// validateInternalExtension validates the given internal extension.
+func validateInternalExtension(extension string) error {
+	if extension == "" {
+		return fmt.Errorf("Extension cannot be empty")
+	}
+
+	// Internal extensions should be in the format `internal:<extension>` with only lowercase letters and underscores.
+	if !internalExtensionRegex.MatchString(extension) {
+		return &invalidInternalExtensionErr{invalidExt: extension}
+	}
+
+	return nil
+}
+
+// HasExtension reports whether the extension set supports the given extension.
+func (e *Extensions) HasExtension(ext string) bool {
+	_, ok := e.external[ext]
+	if ok {
+		return ok
+	}
+
+	_, ok = e.internal[ext]
+	return ok
+}
+
+// Version returns the number of extensions in the set, representing its version number.
+func (e *Extensions) Version() int {
+	return len(e.internal) + len(e.external)
+}
+
+// NewExtensionRegistry creates a new instance of the Extensions struct, which represents a registry of extensions.
+// We have the option to populate the internal extensions or not. Having s option becomes handy
+// when we want to create a custom registry of extensions (see `NewExtensionRegistryFromList`)
+func NewExtensionRegistry(populateInternal bool) (*Extensions, error) {
+	internal := make(map[string]struct{})
+	if populateInternal {
+		// Populate internal extensions here.
+		internal["internal:runtime_extension_v1"] = struct{}{}
```

Comment:
Can we move this to a separate list of strings as in LXD? Then iterate over them here.

Reply:
fixed